packages/pigeon/lib/dart_generator.dart

```diff
@@ -831,17 +811,54 @@ String _getArgumentName(int count, NamedType field) =>
 
 /// Generates the arguments code for [func]
 /// Example: (func, getArgumentName) -> 'String? foo, int bar'
-String _getMethodArgumentsSignature(
-  Method func,
-  String Function(int index, NamedType arg) getArgumentName,
-) {
-  return func.arguments.isEmpty
-      ? ''
-      : indexMap(func.arguments, (int index, NamedType arg) {
-          final String type = _addGenericTypesNullable(arg.type);
-          final String argName = getArgumentName(index, arg);
-          return '$type $argName';
-        }).join(', ');
+String _getMethodArgumentsSignature(Method func) {
+  if (func.arguments.isEmpty) {
+    return '';
+  }
+  bool firstNonPositionalArgument = true;
+  bool firstOptionalPositional = true;
+
+  final List<String> stringArgs = <String>[];
+
+  int index = 0;
+  for (final Parameter arg in func.arguments) {
+    String preArgSymbol = '';
+    if (!arg.isPositional && firstNonPositionalArgument) {
+      firstNonPositionalArgument = false;
+      preArgSymbol = '{';
+    } else if (arg.isPositional && arg.isOptional && firstOptionalPositional) {
+      firstOptionalPositional = false;
+      preArgSymbol = '[';
+    }
+
+    final String required =
+        arg.isRequired && !arg.isPositional ? 'required ' : '';
+
+    final String type = _addGenericTypesNullable(arg.type);
+
+    final String argName = _getArgumentName(index, arg);
+
+    final String defaultValue =
+        arg.defaultValue == null ? '' : ' = ${arg.defaultValue}';
+
+    String postArgSymbol = '';
+    if (!arg.isPositional &&
+        !firstNonPositionalArgument &&
+        func.arguments.length - 1 == index) {
+      postArgSymbol = '}';
+    } else if (arg.isPositional &&
+        !firstOptionalPositional &&
+        (func.arguments.length - 1 == index ||
+            func.arguments[index + 1].defaultValue == null ||
+            !func.arguments[index + 1].isPositional)) {
+      postArgSymbol = ']';
+    }
+    stringArgs
+        .add('$preArgSymbol$required$type $argName$defaultValue$postArgSymbol');
+    index++;
+  }
```

Comment:
[Dart's parameter rules](https://dart.dev/language/functions#parameters) are pretty strict; I'm pretty sure you could replace all of this with something like (freehand, so treat as pseudocode):

```dart
final List<Parameter> requiredPositionalParams = func.arguments.where((Parameter p) => p.isPositional && !p.isOptional);
final List<Parameter> optionalPositionalParams = func.arguments.where((Parameter p) => p.isPositional && p.isOptional);
final List<Parameter> namedParams = func.arguments.where((Parameter p) => !p.isPositional);

String getParameterString(Parameter p) {
  // your required/type/argName/defaultValue logic here
  return 'required$type $argName$defaultValue';
}
String baseParameterString = requiredPositionalParams.map((Parameters p) => getParameterString(p)).join(', ');
String optionalParameterString = optionalPositionalParams.map((Parameters p) => getParameterString(p)).join(', ');
String namedParameterString = namedParams.map((Parameters p) => getParameterString(p)).join(', ');
// Parameter lists can end with either named or optional positional parameters, but not both.
if (namedParameters.isNotEmpty) {
  return '$baseParameterString, {namedParameterString}';
}
if (optionalPositionalParameters.isNotEmpty) {
  return '$baseParameterString, [optionalPositionalParameters]';
}
return baseParameterString;
```

(This assumes we don't actually need `_getArgumentName` here; AFAICT we're only calling this method for host API and Flutter API methods, and AFAIK it's not possible to have nameless parameters in this context. If we do actually need that in some case, I can provide code to pre-map to records that carry the index along.)

This or something like it avoids the need to keep track of a state machine, so it's easier to reason about the method.